feat: enable session leaks prevention by cleaning up long-running tra… (#2655)
googleapis/java-spanner: session-and-channel-pool-configuration.md

1. The `TransactionRunner` will automatically commit the transaction if the supplied user code
   finished without any errors. The `Commit` RPC that is invoked uses a thread from the default gRPC
   thread pool.

### Session Leak

> A Spanner object of the Client Library has a limit on the number of maximum sessions. For example the 

Reviewer: nit:
```suggestion
A `DatabaseClient` object of the Client Library has a limit on the number of maximum sessions. For example the 
```

default value of `MaxSessions` in the Java Client Library is 400. You can configure these values at the time of 

> Client side Database object creation by passing in the `SessionPoolOptions`. When all the sessions are checked 

Reviewer: ```suggestion
creating a `Spanner` instance by setting custom `SessionPoolOptions`. When all the sessions are checked 
```

out of the session pool, every new transaction has to wait until a session is returned to the pool. 
If a session is never returned to the pool (hence causing a session leak), the transactions will have to wait 
indefinitely and your application will be blocked.

#### Common Root Causes

> The most common reason for a session leak is that a transaction was started by the application but 
> never committed or rolled back. What you should do is simply call commit on the transaction at the 
> end of your transaction code. Spanner has two types of transactions, read only and read-write 
> transactions. When we perform a read in a read-write transaction we still need to commit it.

Reviewer: ```suggestion
The most common reason for session leaks in the Java client library are:
1. Not closing a `ResultSet` that is returned by `executeQuery`. Always put `ResultSet` objects in a try-with-resources block, or take other measures to ensure that the `ResultSet` is always closed.
2. Not closing a `ReadOnlyTransaction` when you no longer need it. Always put `ReadOnlyTransaction` objects in a try-with-resources block, or take other measures to ensure that the `ReadOnlyTransaction` is always closed.
3. Not closing a `TransactionManager` when you no longer need it. Always put `TransactionManager` objects in a try-with-resources block, or take other measures to ensure that the `TransactionManager` is always closed.
```

As shown in the example below, the `try-with-resources` block releases the session after it is complete. 
If you don't use `try-with-resources` block, unless you explicitly call the `close()` method on all resources 

> such as `ResultSet` the session is not released back to the pool. If the transaction does not run within 
> `try-with-resources` block and if we don't close the resources explicitly, we will have a session leak.

Reviewer: ```suggestion
such as `ResultSet`, the session is not released back to the pool.
```

```java
DatabaseClient client =
    spanner.getDatabaseClient(DatabaseId.of("my-project", "my-instance", "my-database"));
try (ResultSet resultSet =
    client.singleUse().executeQuery(Statement.of("select col1, col2 from my_table"))) {
  while (resultSet.next()) {
    // use the results.
  }
}
```

#### Debugging and Resolving Session Leaks

##### Logging

> Enabled by default, the logging option shares warn logs when you have exhausted >95% of your session pool. 

Reviewer 2: I think for java we won't have this log right so we can remove the line?

This could mean two things, either you need to increase the max sessions in your session pool (as the number 
of queries run using the client side database object is greater than your session pool can serve) or you may 
have a session leak.

To help debug which transactions may be causing this session leak, the logs will also contain stack traces of 
transactions which have been running longer than expected. The logs are pushed to a destination based on 
how the log exporter is configured for the host application.

> 

Reviewer: Are we logging in the Java client? Based on the comment above, I would expect 'No'.

If we are logging: I'm pretty sure that we are not pushing those logs to any 'destination based on how the log exporter is configured'.

Author: Some of the other client libraries such as Node/Python, emit a log every-time the application uses > 95% of the sessions. In Java, we use the condition as a pre-condition for the algo but don't emit any log for this condition. We only emit a one of the below logs when, 
a) WARN LOG - when there is an actual long-running transaction. This includes bunch of other pre-conditions conditions (lastUsedTime, type of Transaction, etc.)
OR b) WARN LOG - We have automatically cleaned a long running transaction

Now what I mean as pushing log to the exporter is - the logs will be stored in a destination as per what is configured for the application. For ex - Some apps may configure cloud logging while others could configure logs to be stored on separate files on the app server. 

The note about the exporter is not too relevant for Java where there is a default logging behaviour and I think it was more relevant for other languages like Node where there was a logging capability was absent. In Node, the default logging was console logs or you could define some other exporter.

LMK if this comment can be better framed or excluded in context of Java client.

Reviewer: We're using `java.util.logging` for this, right? Then I would just explicitly call out that. So something like 'The warning is logged using java.util.logging.'. I think that most people know what they need to do to configure that, and can otherwise search for it.
